### DISCUSSION TITLE
Can't write type-safe implementations of Tuple methods

### DISCUSSION THREAD

**Discussion by TomasMikula** [scala/scala3#12388] [General Discussion]
The implementations of [`Tuple`](https://github.com/lampepfl/dotty/blob/master/library/src/scala/Tuple.scala) methods delegate to unsafe `runtime.Tuples` methods and then use a type cast (`asInstanceOf`).

[Tuples bring generic programming to Scala 3](https://www.scala-lang.org/2021/02/26/tuples-bring-generic-programming-to-scala-3.html) shows how we could safely implement the `++` (`concat`) method.

However, I'm at a loss to know how to safely implement other methods, such as `drop`, `take`, `splitAt`, to the point that I don't think it's possible (but perhaps just undocumented?).

Below is a failed attempt at implementing the `drop` method. I would be happy to be shown the correct way, but for now I just think it is not possible.

## Compiler version

3.0.0-RC

## Minimized example

```Scala
import compiletime._
import compiletime.ops.int._
import Tuple.Drop

def drop[T <: Tuple, N <: Int](t: T, n: N): Drop[T, N] =
  n match {
    case _ : 0 => t
    case n : S[n1] => t match {
      case _ : EmptyTuple => EmptyTuple
      case (_ *: t1): (_ *: t1) => drop[t1, n1](t1, n-1)
    }
  }
```

## Output

```scala
-- [E007] Type Mismatch Error: drop.scala:9:29 ---------------------------------
9 |      case _ : EmptyTuple => EmptyTuple
  |                             ^^^^^^^^^^
  |Found:    EmptyTuple.type
  |Required: T match {
  |  case EmptyTuple => EmptyTuple
  |  case 
  |    [x, xs <: Tuple] =>> 
  |      scala.runtime.MatchCase[x *: xs, LazyRef(scala.Tuple.Drop[xs, n1])]
  |} <: Tuple
  |
  |where:    T  is a type in method drop with bounds >: (?1 : EmptyTuple.type) and <: Tuple
  |          n1 is a type in method drop with bounds <: Int

longer explanation available when compiling with `-explain`
-- [E007] Type Mismatch Error: drop.scala:10:52 --------------------------------
10 |      case (_ *: t1): (_ *: t1) => drop[t1, n1](t1, n-1)
   |                                                    ^^^
   |                Found:    Int
   |                Required: n1
   |
   |                where:    n1 is a type in method drop with bounds <: Int

longer explanation available when compiling with `-explain`
2 errors found
```

**Comment by odersky** [comment]
Moving to discussions. It would be good if others could participate.

**Comment by nicolasstucki** [comment]
I believe the way to do it is using inline. The issue is that it will be `O(n²)` and will generate a lot of code.

**Reply by TomasMikula** [reply]
> I believe the way to do it is using inline.

Can you elaborate? Just naively adding `inline` to `def drop` has no effect (I get the same 2 errors).

> The issue is that it will be `O(n²)` and will generate a lot of code.

Let's not worry about that for now. (The `O(n²)` would go away if I used a custom recursive tuple implementation. Now I just want to explore what's possible wrt. abstracting over arity.)

**Comment by LPTK** [comment]
The problem is the `n-1` part. It's never going to work unless you somehow partially-evaluate it at compile time, which I believe is what @nicolasstucki was suggesting, and which would probably require a _transparent_ inline method. However, I'm pretty sure this is not what you want, as we don't necessarily know the size as a constant at compile time. Also I don't actually see how this would help type check your definition without having to use casts.

So I think you're going to have much fewer troubles going with the usual way sized vectors are done in dependently-typed languages (look up how it's done in, eg, Idris, if you don't know). Especially now that https://github.com/lampepfl/dotty/pull/12397 is in. The main problem is that forces you to use a custom, inefficient `Nat` data type. We should be able to use `Int`, but for that we'd need a value-level `S` constructor reflecting the type-level `S`. Its absence from the library seems like a big omission.

**Reply by TomasMikula** [reply]
> The problem is the `n-1` part. It's never going to work unless you somehow partially-evaluate it at compile time, which I believe is what @nicolasstucki was suggesting, and which would probably require a _transparent_ inline method. However, I'm pretty sure this is not what you want, as we don't necessarily know the size as a constant at compile time. Also I don't actually see how this would help type check your definition without having to use casts.

That's my sense, too. Blanket statements like _"We can use the same approach for other functions on tuples"_ (from the referenced [article at scala-lang.org](https://www.scala-lang.org/2021/02/26/tuples-bring-generic-programming-to-scala-3.html)) are making a disservice to the community by creating false expectations.

> So I think you're going to have much fewer troubles going with the usual way sized vectors are done in dependently-typed languages

I don't think a reference to how it would be done in a dependently typed language is useful, because it cannot be replicated without dependent types 🙂 

I would divide the potential approaches into two categories, based on how the return type is declared:

1. The return type is a member type of some (implicit) evidence argument:

    ```scala
    def drop[T <: Tuple, N <: Nat](t: T, n: N)(implicit ev: Drop[T, N]): ev.Out
    ```

    We know this is feasible, although cumbersome, even prior to Scala 3.

    The implementation of this method will just delegate to a method on the evidence argument. (In an ideal world, it could perhaps be implemented by pattern matching on `t` and `ev` simultaneously (assuming a suitable pattern-matchable representation of `Drop[T, N]`), but I don't have that much confidence in Scala's type inference in pattern matching.)

I was wondering whether match types in Scala 3 offer an alternative and less cumbersome solution (as the mentioned article would suggest). 

2. The return type is a function on types (a match type):

    ```scala
    def drop[T <: Tuple, N <: Int](t: T, n: N): Drop[T, N]
    ```

    This doesn't seem to be workable at the moment. At least not with `Int`s (and thus also not with the built-in tuples). I might give it a try with custom `Nat` and `Tuple` types (but my hopes wrt. pattern matching type inference are again low).

    _What I would like someone involved to clarify is whether this is even a use case that's ultimately supposed to be supported by match types. And I mean supported **intentionally,** as in the feature is not complete without it._

**Reply by LPTK** [reply]
> I don't think a reference to how it would be done in a dependently typed language is useful, because it cannot be replicated without dependent types 🙂

It's funny that people have this impression that Scala is not a dependently typed language. It totally is. All you need is the ability of referring to unknown types arising from values (path-dependent types do that for you), and the ability of expressing equality proofs (abstract type member bounds can do that for you).

This ends up being quite verbose, but the new GADT reasoning logic, which is [based on these fundamental principles](https://dl.acm.org/doi/abs/10.1145/3337932.3338813), makes the whole thing a bit more bearable. Here is an example of a sized vector in Scala 3:

```scala
enum Nat[Self]:
  case Z() extends Nat[Z]
  case S[N](prev: Nat[N]) extends Nat[S[N]]

import Nat._

enum Vec[N, +T]:
  case Cons[M, T](head: T, tail: Vec[M, T]) extends Vec[S[M], T]
  case Nil() extends Vec[Z, Nothing]

def mkVec[N, T](n: Nat[N], e: T): Vec[N, T] = n match
  case Z() => Vec.Nil()
  case S(n) => Vec.Cons(e, mkVec(n, e))

@main def test =
  type _3 = S[S[S[Z]]]
  val _3 = S(S(S(Z())))
  val v = mkVec(_3, "hi")
  val w: Vec[_3, String] = v
  println(w) // Cons(hi,Cons(hi,Cons(hi,Nil())))
```

https://scastie.scala-lang.org/FZ828pjQRBi9NeYMpb1RMg

**Reply by TomasMikula** [reply]
You can take a loose definition of dependent types so that Scala fits it, but that doesn't change that the implementation of operations on heterogeneous lists (or even just sized vectors) does not translate in a straightforward way from a dependently typed language (in the common narrow sense) to Scala. The important missing part on the Scala side (which btw. match types try to emulate somewhat) is the ability for types to be defined by arbitrary functions (whose body is transparent), as in

```scala
def foo(a: A, b: bar(a)): baz(a, b)

// where

def bar(a: A): Type            = { /* arbitrary code returning a type */ }
def baz(a: A, b: bar(a)): Type = { /* arbitrary code returning a type */ }
```

And yes, GADTs go a long way as a substitute, if instead of the function `bar(A)` one defines a GADT `Bar[T]` whose structure mimics the pattern matching structure of the function `bar`, and similarly for replacing `baz` by a GADT `Baz`. However, for that you need GADT pattern matching to _really_ work, and Scala is notoriously bad at that. (I am aware that there are significant improvements in Scala 3, but I still don't trust that it will not let me down after having invested some non-trivial effort, because toy examples seemed to work. Yes, I have trust issues, because I have been let down so many times.)

Still, I wouldn't call the GADT encoding _"the usual way [things] are done in dependently-typed languages"._ A dependently typed language might not even have GADTs (and does not need them).

**Reply by TomasMikula** [reply]
> for that you need GADT pattern matching to _really_ work

Here's one GADT problem I bumped into when playing with GADT encoding of dependent types: #12478.

**Reply by LPTK** [reply]
> You can take a loose definition of dependent types so that Scala fits it

That's actually quite a narrow definition! People are so dogmatic these days. No, "dependent types" does not mean "essentially Idris". In fact, there is a wealth of dependent type systems, many much farther away from Idris than Scala itself is.

> the implementation of operations on heterogeneous lists (or even just sized vectors) does not translate in a straightforward way from a dependently typed language (in the common narrow sense) to Scala

I think it actually does, once you figure out the way the two encodings correspond to each other. But granted, current bugs and limitations in the current Scala compiler implementation tend to stand in the way of that.

> The important missing part on the Scala side (which btw. match types try to emulate somewhat) is the ability for types to be defined by arbitrary functions (whose body is transparent)

That's actually not missing. As you point out, match types can fill this role. And before that, in Scala 2, type projection could do it. Sure, it's not particularly _convenient_, but unifying term and type functions is definitely not _required_.

> there are significant improvements in Scala 3, but I still don't trust that it will not let me down

Not an unfounded fear IMHO. Take a look at this issue: https://github.com/lampepfl/dotty/issues/8900

The main problem is the tricky interactions between type inference and the new GADT reasoning. But note that in terms of foundations, the Dependent Object Type calculus (DOT) on which Scala 3's type system is conceptually built already has all the tools to do dependently-typed programming. GADT reasoning and type inference are just "sugar" built on top of it.

> A dependently typed language might not even have GADTs (and does not need them).

I disagree, given "dependent type" in your narrower sense. Dependent type theories in the line going back to Martin Löf included the notion of [type equality/identity, which is essentially a "proto inductive type"](https://ncatlab.org/nlab/show/identity+type#Definition).

**Reply by TomasMikula** [reply]
> That's actually not missing. As you point out, match types can fill this role. And before that, in Scala 2, type projection could do it. Sure, it's not particularly _convenient_, but unifying term and type functions is definitely not _required_.

Match types are the point of investigation of this discussion. This thread though digresses into an alternative approach. It is definitely not convenient. One quickly runs into problems like: given `a, b: Drop[T, N]`, the compiler doesn't know that `a.Out = b.Out`, i.e. it doesn't know that `Drop[T, N]` is something like a function.

> I disagree, given "dependent type" in your narrower sense. Dependent type theories in the line going back to Martin Löf included the notion of [type equality/identity, which is essentially a "proto inductive type"](https://ncatlab.org/nlab/show/identity+type#Definition).

Let me clarify. What I meant by _"a dependently typed language not needing GADTs"_ was just that it can encode a GADT using a plain ADT (an enumeration of cases/indices) and a function from that ADT to `Type`.

**Reply by LPTK** [reply]
> it can encode a GADT using a plain ADT (an enumeration of cases/indices) and a function from that ADT to `Type`

Can it? How do you encode `Eq a b`, so as to write an implementation for `forall a b. Eq a b -> a -> b`?

**Reply by TomasMikula** [reply]
> > it can encode a GADT using a plain ADT (an enumeration of cases/indices) and a function from that ADT to `Type`
> 
> Can it? How do you encode `Eq a b`, so as to write an implementation for `forall a b. Eq a b -> a -> b`?

OK, and a built-in equality type :)

**Reply by LPTK** [reply]
Actually, you'd still be missing some form of existential quantification. It's an essential ingredient of modern "dependently-typed" languages, supported for instance through dependent pairs, inductive types, or GADTs. This includes Scala – which supports it through object types with abstract type members.

**Reply by TomasMikula** [reply]
Yeah, I was kind of implicitly assuming existential quantification. It is actually already at the very core of the imagined encoding of a GADT. It would be `Σ (i: I) f(i)` where `f: I -> Type` and `I` is a simple enumeration. And `f` might need to return a dependent pair as well.
